### DISCUSSION TITLE
[WIKI] `XSharedPreferences` minimum sdk level `>27`

### DISCUSSION THREAD

**Discussion by programminghoch10** [LSPosed/LSPosed#765] [Q&A] [answered]
The wiki about the new `XSharedPreferences` says the minimum api level is `93`, 
so as long as the api level is `93` the new `XSharedPreferences` should be available.
LSPosed has a minimum sdk level of `27`. 
The api level `93` is supported on sdk level `27`.

Still the wiki about new `XSharedPreferences` says the sdk level for that feature must be `>27` so at least `28`.
Is this intentional? I don't see why api `93` should be fully supported on sdk `27`, 
but new `XSharedPreferences` (available on api `93`) needs sdk `>27`.

I guess it's just a typo.
If so please correct the wiki to state sdk `>=27`.

**Comment by yujincheng08** [comment, marked as answer]
It means if ur module targert sdk 27+, you will need this feature. Its not a min requirement if this feature.

**Reply by programminghoch10** [reply]
yeah but its available when api is `93` on an android 8.1 device
why does it say sdk `>27` and not sdk `27+` then

**Reply by yujincheng08** [reply]
?
when targeting sdk 27 u dont need this featrue.

**Reply by programminghoch10** [reply]
but can i use it on sdk 27? i dont want to implement a different solution just for sdk 27
question is, does the new `XSharedPreferences` also work on sdk 27?

**Reply by yujincheng08** [reply]
we never said this. you misunderstood it.

**Reply by programminghoch10** [reply]
Well I have an app targeting sdk 29, which uses the new `XSharedPreferences`.
can i just set the minimum sdk level down to `27` and expect it to work on an android 8.1 device?

**Reply by yujincheng08** [reply]
why not

**Reply by programminghoch10** [reply]
so because its targeted for sdk 29, which is >27 the new xsharedpreferences will also be available on an sdk 27 device
okay i think i get it now, kinda confusing tbh

**Reply by yujincheng08** [reply]
?
even if you target 27, you can still use this feature. But you DONT HAVE TO use it since the original one works. Understand?

**Reply by programminghoch10** [reply]
Yes thank you, got it now.
One more question if you don't mind...
Why does the original one work on android 8.1?
Looking through the source code [`checkMode`](https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/app/ContextImpl.java#2868) does throw SecurityExceptions from android 7 onwards.

**Reply by yujincheng08** [reply]
We hooked `checkMode` if this feature is on.

**Reply by programminghoch10** [reply]
Do we always hook `checkMode` even when sdk=27 and api<93 and the new `XSharedPreferences` is disabled?

**Reply by yujincheng08** [reply]
No. we only hook it when this one is enabled.

**Reply by programminghoch10** [reply]
but then I do need to use the new `XSharedPreferences` on sdk 27 because even the original implementation would not work when `checkMode` is not hooked

**Reply by yujincheng08** [reply]
it works as soon as u manually set file to world readable. This is how devs did before this feature is released.

**Reply by programminghoch10** [reply]
so using the new implementation even on sdk 27 is possible and easier than to manually set the files permissions
alright thank you
